bothan-api/server/src/api/crypto.rs

```diff
@@ -0,0 +1,134 @@
+            Err(SetRegistryError::InvalidRegistry) => {
+                Ok(registry_resp(UpdateStatusCode::InvalidRegistry))
+            }
+            Err(SetRegistryError::FailedToRetrieve(_)) => {
+                Ok(registry_resp(UpdateStatusCode::FailedToGetRegistry))
+            }
+            Err(SetRegistryError::UnsupportedVersion) => {
+                Ok(registry_resp(UpdateStatusCode::UnsupportedVersion))
+            }
+            Err(SetRegistryError::FailedToParse) => {
+                error!("Failed to parse registry");
+                Err(Status::invalid_argument(
+                    "Registry is incorrectly formatted",
+                ))
+            }
+            Err(SetRegistryError::InvalidHash) => {
+                error!("Invalid IPFS hash");
+                Err(Status::invalid_argument("Invalid IPFS hash"))
+            }
+            Err(SetRegistryError::FailedToSetRegistry(_)) => {
+                error!("Failed to set registry");
+                Err(Status::internal("Failed to set registry"))
+            }
+        }
+    }
+
+    async fn set_active_signal_id(
+        &self,
+        request: Request<SetActiveSignalIdRequest>,
+    ) -> Result<Response<SetActiveSignalIdResponse>, Status> {
+        info!("received set active signal id request");
+        let update_registry_request = request.into_inner();
+        let mut manager = self.manager.write().await;
+        let set_result = manager
+            .set_active_signal_ids(update_registry_request.signal_ids)
+            .await;
+
+        match set_result {
+            Ok(_) => {
+                info!("successfully set active signal ids");
+                Ok(Response::new(SetActiveSignalIdResponse { success: true }))
+            }
+            Err(e) => {
+                error!("failed to set active signal ids: {}", e);
+                Ok(Response::new(SetActiveSignalIdResponse { success: false }))
+            }
+        }
+    }
+
+    async fn get_price(
+        &self,
+        request: Request<PriceRequest>,
+    ) -> Result<Response<PriceResponse>, Status> {
+        info!("received get price request");
+        let price_request = request.into_inner();
+        let manager = self.manager.read().await;
+        let price_states = manager.get_prices(price_request.signal_ids.clone()).await;
+
+        let prices = price_request
+            .signal_ids
+            .into_iter()
+            .zip(price_states)
+            .map(|(id, state)| match state {
+                PriceState::Available(raw_price) => {
+                    let mantissa_price = raw_price
+                        .round_dp_with_strategy(PRECISION, RoundingStrategy::ToZero)
+                        .mantissa();
+                    match i64::try_from(mantissa_price) {
+                        Ok(p) => price(&id, PriceStatus::Available, p),
+                        Err(_) => price(&id, PriceStatus::Unavailable, 0),
```

Comment:
do we expect this to happen? should we log the error out? i'm not sure it should be info, warn, or error.